File: 19-03-2024-Task-Scheduler.py
class Solution:
    def leastInterval(self, tasks, n):
        size = len(tasks)
        mp = {}
        for it in tasks:
            if it in mp:
                mp[it] += 1
            else:
                mp[it] = 1
        maxHeap = []
        for it in mp:
            maxHeap.append(mp[it])
        maxHeap.sort(reverse=True)
        
        q = []
        time = 0
        while len(maxHeap) > 0 or len(q) > 0:
            time += 1
            if len(maxHeap) > 0:
                curr = maxHeap[0]
                maxHeap.pop(0)
                count = curr - 1
                if count > 0:
                    q.append((count, time + n))
            if len(q) > 0 and q[0][1] == time:
                maxHeap.append(q[0][0])
                q.pop(0)
        return time

File: 19-03-2024-Task-Scheduler.java
import java.util.*;

class Solution {
    public int leastInterval(List<Character> tasks, int n) {
        int size = tasks.size();
        Map<Character, Integer> mp = new HashMap<>();
        for (char it : tasks) {
            mp.put(it, mp.getOrDefault(it, 0) + 1);
        }
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
        for (int it : mp.values()) {
            maxHeap.offer(it);
        }
        
        Queue<Pair<Integer, Integer>> q = new LinkedList<>();
        int time = 0;
        while (!maxHeap.isEmpty() || !q.isEmpty()) {
            time++;
            if (maxHeap.size() > 0) {
                int curr = maxHeap.poll();
                int count = curr - 1;
                if (count > 0) {
                    q.offer(new Pair<>(count, time + n));
                }
            }
            if (!q.isEmpty() && q.peek().getValue() == time) {
                maxHeap.offer(q.peek().getKey());
                q.poll();
            }
        }
        return time;
    }
}

